src/Analyser/Analyser.php

```diff
@@ -104,6 +109,42 @@ public function analyse(
 			$postFileCallback(1);
 		}
 
+		// verify serializability of collected data
+		$alreadyReported = [];
+		foreach ($collectedData as $collected) {
+			if (array_key_exists($collected->getCollectorType(), $alreadyReported)) {
+				continue;
+			}
+
+			$json = json_encode($collected->getData());
+			if ($json === false) {
+				$errors[] = $this->buildCollectedDataError(
+					sprintf(
+						'Data collected by %s is not JSON serializable.',
+						$collected->getCollectorType(),
+					),
+					$collected,
+				);
+				$alreadyReported[$collected->getCollectorType()] = true;
+
+				break;
+			}
+
+			$restored = json_decode($json);
+			if ($restored === $collected->getData()) {
```

Comment:
thinking about this, is this actually the correct check?

if there is an object that implements JsonSerializable returned from a collector it will encode fine, but when decoded it will be an `stdClass` meaning the `===` check would fail

if it's intended that you can use JsonSerializable in collectors then the check needs to be a bit more robust

Reply:
you describe exactly the case, why this rule was created. the behaviour of phpstan is different depending on whether phpstan is running with parallel worker or with a single worker (e.g. in `--debug`), since the collected data is only json encoded in the the worker and decoded in the main thread in parallel, which means when you put a `JsonSerializable` object into collected data it will work differently depending on the worker scenario

Reply:
ah my misunderstanding, when Ondrej said 

> This type would not be precise. Anything that is JSON-serializable works.

i thought the intention was that JsonSerializable should be permitted, thanks for clearing that up